torch/distributed/_spmd/api.py

```diff
@@ -306,6 +318,43 @@ def _fused_adam_decomp(
 }
 
 
+DEDUP_TARGETS: Set[torch._ops.OpOverload] = {
+    aten.all_reduce.default,
+    aten.wait_tensor.default,
+}
+
+
+def _dedup_collectives(gm: fx.GraphModule) -> fx.GraphModule:
+    # deduplicate collectives
+    node_to_node: Dict[fx.Node, fx.Node] = {}
+    args_to_node: Dict[Tuple[Any, ...], fx.Node] = {}
+
+    nodes_to_erase: List[fx.Node] = []
+
+    for node in gm.graph.nodes:
+        # replace all args with the results from the first unique comm op
+        args, spec = pytree.tree_flatten(node.args)
+        unique_args = [node_to_node.get(a, a) for a in args]
+        node.args = pytree.tree_unflatten(unique_args, spec)
+
+        if node.target in DEDUP_TARGETS:
+            args_key = (node.target, *unique_args)
+            unique_node = args_to_node.get(args_key, None)
+            if unique_node is None:
+                # first time seeing this combination, remember it
+                args_to_node[args_key] = node
+            else:
+                # the current node is a duplicate, replace it
+                node_to_node[node] = unique_node
+                nodes_to_erase.append(node)
+
+    # erase all duplicated nodes
+    for node in nodes_to_erase:
+        gm.graph.erase_node(node)
```

Comment:
don't we need recompile the gm to make sure the dedup takes into effect?

Reply:
yep, good catch!